### PR TITLE
[native dsl] Add Helion cross entropy override for B200

### PR DESCRIPTION
**Pull request by karthickai**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #190300

Building on the Helion backend registration ([#190636](https://github.com/pytorch/pytorch/pull/190636)), add an SM100 cross-entropy override using pretuned AOT configurations. The 21 pretuned shapes map to seven fixed configurations without runtime autotuning. Helion is enabled for 19 shapes; 2048x32000 and 4096x32000 remain on ATen because validation and launch overhead prevent reliable latency improvements.

Eligible inputs are contiguous and aligned BF16 logits with int64 class-index targets, mean reduction, default `ignore_index=-100`, no weight or label smoothing, and no autograd or autocast. Label validation and nonignored-count calculation remain asynchronous on-device. The log-sum-exp and final reduction use FP32 accumulation before casting the scalar result to the eager output dtype. Eligible COW inputs use read-only Helion launches without materializing the tensors.

Unsupported contracts, requires-grad and forward-AD inputs, autocast, compile/export, CUDA Graph capture, and tensor subclasses retain ATen semantics. If Helion is unavailable, incompatible, disabled through the native-JIT setting, or using an unsupported backend, no cross-entropy override or associated dispatch-key registrations are installed.

Across the 19 Helion-enabled shapes, the measured GB200 geomean speedup is 1.687x.

GB200 BF16 CUDA-event latency. Each row reports the median of six measurements per path. Each measurement uses 25 ms warmup and 100 ms timing. ATen/native ordering alternates deterministically within the six paired batches. Speedup is ATen median divided by native-path median.

GPU: NVIDIA GB200 | torch: 2.15.0a0+gitd090b56 | Helion: 1.4.1.dev134+g188c6f08c | backend: triton

| Shape | ATen (us) | Native path (us) | Speedup |
|---:|---:|---:|---:|
| 2048x32000 | 134.18 | 134.18 | 1.000x (ATen fallback) |
| 4096x32000 | 232.59 | 232.54 | 1.000x (ATen fallback) |
| 8192x32000 | 427.06 | 240.64 | 1.775x |
| 8192x128000 | 1231.89 | 684.03 | 1.801x |
| 16384x128000 | 2417.68 | 1273.87 | 1.898x |
| 32768x128000 | 4800.50 | 2467.87 | 1.945x |
| 2048x128256 | 345.73 | 256.03 | 1.350x |
| 4096x128256 | 644.11 | 389.91 | 1.652x |
| 8192x128256 | 1238.56 | 702.46 | 1.763x |
| 16384x128256 | 2429.93 | 1319.94 | 1.841x |
| 2048x129280 | 353.31 | 236.02 | 1.497x |
| 4096x129280 | 657.42 | 397.34 | 1.655x |
| 8192x129280 | 1267.71 | 750.50 | 1.689x |
| 2048x151936 | 427.04 | 301.92 | 1.414x |
| 4096x151936 | 804.88 | 526.32 | 1.529x |
| 8192x151936 | 1568.27 | 995.34 | 1.576x |
| 2048x152064 | 426.34 | 253.30 | 1.683x |
| 4096x152064 | 801.84 | 420.61 | 1.906x |
| 8192x152064 | 1565.68 | 821.28 | 1.906x |
| 1024x256000 | 373.79 | 251.54 | 1.486x |
| 2048x256000 | 694.30 | 372.77 | 1.863x |
| **Helion-enabled geomean (19 shapes)** | | | **1.687x** |

### CODE REVIEW

File: torch/_native/helion_utils.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
+import functools
+import logging
```
Comment (slayton58) [inline torch/_native/helion_utils.py:2]: All the fundamental adding of helion as a backend should be in a separate commit in the stack

File: torch/_native/helion_utils.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+def _check_runtime_available() -> tuple[bool, Version | None]:
+    if not _cuda.is_built():
+        return (False, None)
+
+    import torch
+
+    if torch.version.hip is not None:
+        return (False, None)
+
+    reason = _unavailable_reason([("helion", "helion")])
```
Comment (slayton58) [inline torch/_native/helion_utils.py]: Do you need to make sure there's an appropriate backend (triton/cuteDSL/whatever) to helion as well?

File: torch/_native/registry.py [hunk 3]
```diff
@@ -163,6 +163,13 @@ def __str__(self) -> str:
 # and consumers control exactly when and where overrides take effect.
 _native_decomp_overrides: dict[object, Callable] = {}
 
+# Native backend kernels captured before router installation. Auxiliary
```
Comment (slayton58) [inline torch/_native/registry.py]: What are these needed for?

**Review by slayton58 — COMMENTED** [review]
I know this is a draft, but flagging up some big-picture stuff early

File: torch/_native/ops/cross_entropy/_helion_aot_helion_kernel_cuda_sm100.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+
+def key_cross_entropy(*args) -> int:
+    is_tensor = len(args) > 0 and isinstance(args[0], torch.Tensor)
+    _arg0_dim0 = int(args[0].shape[0]) if is_tensor and args[0].ndim > 0 else 0
+    _arg0_dim1 = int(args[0].shape[1]) if is_tensor and args[0].ndim > 1 else 0
+    _arg0_numel = int(args[0].numel()) if is_tensor else 0
+    if _arg0_numel <= 262144000.0:
+        if _arg0_dim0 <= 1024.0:
+            return 6
```
Comment (slayton58) [inline torch/_native/ops/cross_entropy/_helion_aot_helion_kernel_cuda_sm100.py]: Return some meaningful result instead of an opaque `int`?

File: torch/_native/ops/cross_entropy/helion_impl.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    if self.requires_grad or target.requires_grad:
+        return False
+    if torch.autograd.forward_ad.unpack_dual(self).tangent is not None:
+        return False
+    if torch.is_autocast_enabled("cuda"):
+        return False
+    # Target validation is data-dependent, so this override is eager-only.
+    if type(self) is not torch.Tensor or type(target) is not torch.Tensor:
+        return False
+    is_cow = torch._C._is_cow_tensor  # pyrefly: ignore[missing-attribute]
```
Comment (slayton58) [inline torch/_native/ops/cross_entropy/helion_impl.py]: Use `ReadOnlyTensorWrapper` from https://github.com/pytorch/pytorch/pull/188554 and get rid of COW checks

File: torch/_native/ops/cross_entropy/helion_impl.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    if is_cow(self) or is_cow(target):
+        return False
+    if self.data_ptr() % 16 != 0 or target.data_ptr() % 16 != 0:
```
Comment (slayton58) [inline torch/_native/ops/cross_entropy/helion_impl.py]: Variable instead of magic number

File: torch/_native/ops/cross_entropy/helion_impl.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+    if _autocast_lib is not None:
+        _autocast_lib._destroy()
+        _autocast_lib = None
+    _autograd_lib._destroy()
+    _autograd_lib = None
+    _autocast_fallback_kernel = None
+
+
+def register_to_dispatch() -> None:
+    if (
```
Comment (slayton58) [inline torch/_native/ops/cross_entropy/helion_impl.py]: These should be covered in the `hu.register_op_override` call

File: torch/_native/ops/cross_entropy/helion_impl.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+def _uninstall_autograd_fallthrough() -> None:
+    global _autocast_fallback_kernel, _autocast_lib, _autograd_lib
+    if _autograd_lib is None:
+        return
+    if _autocast_lib is not None:
+        _autocast_lib._destroy()
+        _autocast_lib = None
+    _autograd_lib._destroy()
+    _autograd_lib = None
+    _autocast_fallback_kernel = None
```
Comment (slayton58) [inline torch/_native/ops/cross_entropy/helion_impl.py]: Why do we need these? You're overriding the `cross_entropy_loss` symbol only, everything else (autograd, etc.) shouldn't need touching.

File: torch/_native/helion_utils.py (hunk 1 above)
Comment (karthickai) [reply torch/_native/helion_utils.py:2]: thanks Simon for reviewing early, I split the stack now helion backend integration is here https://github.com/pytorch/pytorch/pull/190636

File: torch/_native/helion_utils.py (hunk 2 above)
Comment (karthickai) [reply torch/_native/helion_utils.py]: I added a  supported-backends gate (_HELION_BACKENDS = ("triton",), selectable via HELION_BACKEND). extending to another backend later is just adding it to _HELION_BACKENDS.

File: torch/_native/ops/cross_entropy/helion_impl.py (hunk 7 above)
Comment (karthickai) [reply torch/_native/ops/cross_entropy/helion_impl.py]: I updated the register_to_dispatch now just calls hu.register_op_override

File: torch/_native/ops/cross_entropy/helion_impl.py (hunk 8 above)
Comment (karthickai) [reply torch/_native/ops/cross_entropy/helion_impl.py]: You're right for a normal op it override the kernel and autograd/autocast wrap it from above. The catch is cross_entropy_loss is CompositeImplicitAutograd installing our CUDA override drops the composite off the Autograd key, which breaks  backward for the whole symbol, not just our inference shapes. So the fallthrough just restores autograd so backward decomposes through aten. AutocastCUDA one is the same idea for a different path without it, torch.compile + autocast silently computes the loss in bf16 instead of aten's fp32.

File: torch/_native/registry.py (hunk 3 above)
Comment (karthickai) [reply torch/_native/registry.py]: I removed it _fallback_kernels cache is gone, the native kernel is captured locally now.

File: torch/_native/ops/cross_entropy/_helion_aot_helion_kernel_cuda_sm100.py (hunk 4 above)
Comment (karthickai) [reply torch/_native/ops/cross_entropy/_helion_aot_helion_kernel_cuda_sm100.py]: I've udpated the selector is select_cross_entropy_config(configs, *args) -> dict and returns the actual config dict instead of a bare index.

File: torch/_native/ops/cross_entropy/helion_impl.py (hunk 6 above)
Comment (karthickai) [reply torch/_native/ops/cross_entropy/helion_impl.py]: it's the kernel's required 16-byte input alignment updated with _REQUIRED_ALIGNMENT.

File: torch/_native/ops/cross_entropy/helion_impl.py (hunk 5 above)
Comment (karthickai) [reply torch/_native/ops/cross_entropy/helion_impl.py]: ReadOnlyTensorWrapper is DLPack-export-only it raises on everything except `__dlpack__`, so it can't be passed to a Helion/Triton launch (which reads .data_ptr()/.dtype/.stride()). I use ConstTensorWrapper instead for the same zero-copy read, matching bmm_outer_product already does.
